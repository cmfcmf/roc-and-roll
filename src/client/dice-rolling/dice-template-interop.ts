import { DiceRollTree } from "../../shared/dice-roll-tree-types-and-validation";
import { RRDiceTemplate } from "../../shared/validation";
import { RRDiceTemplatePart } from "../../shared/state";
import { rrid } from "../../shared/util";

export function tryConvertDiceRollTreeToDiceTemplateParts(
  diceRollTree: DiceRollTree<false>
): RRDiceTemplate["parts"] | false {
  if (diceRollTree.type !== "term" || diceRollTree.operator !== "+") {
    return false;
  }

  const parts: RRDiceTemplate["parts"] = [];
  for (const operand of diceRollTree.operands) {
    switch (operand.type) {
      case "num":
        parts.push({
          id: rrid<RRDiceTemplatePart>(),
          type: "modifier",
          number: operand.value,
          damage: operand.damage,
        });
        break;
      case "dice":
        parts.push({
          id: rrid<RRDiceTemplatePart>(),
          type: "dice",
          count: operand.count,
          faces: operand.faces,
          modified: operand.modified,
          negated: false,
          damage: operand.damage,
        });
        break;
      default:
        return false;
    }
  }

  return parts;
}
