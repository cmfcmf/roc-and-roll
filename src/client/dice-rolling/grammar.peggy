// Grammar written in the peggy format: https://github.com/peggyjs/peggy
//
// Test the grammar here:
// https://peggyjs.org/online.html

Expression
  = head:Term tail:(_ ("+" / "-") _ Term)* {
      return tail.reduce(function(result, element) {
        const operator = element[1];
        return {
          type: 'term',
          operator,
          operands: result.type === 'term' && result.operator === operator
          	? [...result.operands, element[3]]
            : [result, element[3]]
        };
      }, head);
    }

Term
  = head:Factor tail:(_ ("*" / "×" / "/" / "÷") _ Factor)* {
      return tail.reduce(function(result, element) {
        const operator = element[1] === "×"
            ? "*"
            : element[1] === "÷"
            ? "/"
            : element[1];

        return {
          type: 'term',
          operator,
          operands: result.type === 'term' && result.operator === operator
          	? [...result.operands, element[3]]
            : [result, element[3]]
        };
      }, head);
    }

Factor
  = Parens
  / Dice
  / Num
  / Negated

Parens
  = "(" _ expr:Expression _ ")" {
  	return {
      type: 'parens',
      inner: expr,
    };
  }

Dice
  = _ countDigits:([0-9]+)? modifierChar:([dai]) faces:([0-9]+) _ {
    const count = countDigits === null ? 1 : parseInt(countDigits.join(''), 10);
    const modified = {
      	d: 'none',
        a: 'advantage',
        i: 'disadvantage',
      }[modifierChar];

    return {
      type: 'dice',
      count: count === 1 && modified !== "none" ? 2 : count,
      results: 'not-yet-rolled',
      modified,
      faces: parseInt(faces.join(''), 10),
      damage: {
        type: null,
      },
    };
  }

Num
  = (_ "+")? _ [0-9]+("."[0-9]+)? _ {
    return {
  	  type: 'num',
      value: parseFloat(text(), 10),
      damage: {
        type: null,
      },
  	};
  }

Negated
  = sign:("-" _) factor:Factor {
  	return {
      type: 'negated',
      inner: factor,
    }
  }

_ "whitespace"
  = [ \t\n\r]*
