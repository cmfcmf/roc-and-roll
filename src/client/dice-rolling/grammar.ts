import {
  DiceRollTree,
  DRTPartDice,
  DRTPartExpression,
  DRTPartNegated,
  DRTPartNum,
  DRTPartParens,
  DRTPartTerm,
  isDiceRollTree,
} from "../../shared/dice-roll-tree-types-and-validation";
import { assertNever } from "../../shared/util";
import parser from "./grammar.peggy";

export const parseDiceString = (input: string): DiceRollTree<false> => {
  const diceRollTree = parser.parse(input);
  // if (process.env.NODE_ENV !== "production") {
  const errors: string[] = [];
  if (!isDiceRollTree(false)(diceRollTree, { errors })) {
    console.error(diceRollTree);
    console.error(errors);
    throw new Error(
      `The grammar returned an invalid tree. This should never happen.`
    );
  }
  // }

  return diceRollTree;
};

export const parseDiceStringGetSyntaxError = (
  input: string
): InstanceType<typeof parser.SyntaxError> | null => {
  try {
    parser.parse(input);
  } catch (error) {
    if (error instanceof parser.SyntaxError) {
      return error;
    }
    // If any other error occurs, throw it (should never happen).
    throw error;
  }
  return null;
};

// Base class for dice roll tree visitors that can be used to visit each tree
// node to iteratively create a result T. The second template parameter
// indicates whether the visitor handles a dice roll tree with or without the
// dice already having been rolled.
export abstract class DRTVisitor<T, WithResults extends boolean, D = never> {
  public visit(expression: DRTPartExpression<WithResults>, data?: D): T {
    switch (expression.type) {
      case "num":
        return this.visitNum(expression, data);
      case "dice":
        return this.visitDice(expression, data);
      case "term":
        return this.visitTerm(expression, data);
      case "parens":
        return this.visitParens(expression, data);
      case "negated":
        return this.visitNegated(expression, data);
      default:
        assertNever(expression);
    }
  }

  protected abstract visitNum(expression: DRTPartNum, data?: D): T;
  protected abstract visitDice(
    expression: DRTPartDice<WithResults>,
    data?: D
  ): T;
  protected abstract visitTerm(
    expression: DRTPartTerm<WithResults>,
    data?: D
  ): T;
  protected abstract visitParens(
    expression: DRTPartParens<WithResults>,
    data?: D
  ): T;
  protected abstract visitNegated(
    expression: DRTPartNegated<WithResults>,
    data?: D
  ): T;
}
