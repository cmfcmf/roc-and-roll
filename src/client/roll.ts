import { randomBetweenInclusive } from "../shared/random";
import {
  RRMultipleRoll,
  RRLogEntryDiceRoll,
  RRDamageType,
  RRDice,
  RRPlayerID,
  RRModifier,
} from "../shared/state";
import { assertNever } from "../shared/util";

export function rollInitiative(
  mod: number,
  multiple: RRMultipleRoll,
  playerId: RRPlayerID
): Omit<RRLogEntryDiceRoll, "id" | "type" | "timestamp"> {
  return {
    payload: {
      rollType: "initiative",
      dice: [rollD20(multiple), ...(mod === 0 ? [] : [modifier(mod)])],
    },
    silent: false,
    playerId,
  };
}

function modifier(
  modifier: number,
  damageType: RRDamageType = null
): RRModifier {
  return {
    type: "modifier",
    modifier,
    damageType,
  };
}

function rollD20(multiple: RRMultipleRoll = "none") {
  return roll({
    faces: 20,
    count: multiple == "none" ? 1 : 2,
    modified: multiple,
  });
}

export function roll({
  faces,
  count,
  damageType,
  modified,
  negated,
}: {
  faces: number;
  count: number;
  damageType?: RRDamageType;
  modified?: RRMultipleRoll;
  negated?: boolean;
}): RRDice {
  if (modified !== "none" && count <= 1) {
    count = 2;
  }
  const results = Array.from({ length: count }, () =>
    randomBetweenInclusive(1, faces)
  );
  return {
    type: "dice",
    faces,
    modified: modified ?? "none",
    diceResults: results,
    damageType: damageType ?? null,
    negated: negated ?? false,
  };
}

export function diceResultString(logEntry: RRLogEntryDiceRoll) {
  return logEntry.payload.dice
    .map((part) => {
      if (part.type === "modifier") {
        return part.modifier;
      }
      const prefix = part.negated ? "-" : "";
      if (part.type === "dice") {
        if (part.diceResults.length === 1) {
          return `${prefix}${part.diceResults[0]!} (d${part.faces})`;
        }
        if (part.modified === "none") {
          return `${prefix}${part.diceResults
            .map((r) => `${r} (d${part.faces})`)
            .join(" + ")}`;
        }
        return `${prefix}(${part.diceResults
          .map((r) => `${r} (d${part.faces})`)
          .join(", ")})`;
      }
      assertNever(part);
    })
    .join(" + ");
}

export function diceResult(logEntry: RRLogEntryDiceRoll) {
  return logEntry.payload.dice
    .map((part) => {
      if (part.type === "modifier") {
        return part.modifier;
      }
      if (part.type === "dice") {
        const sign = part.negated ? -1 : 1;
        if (part.modified === "none") {
          return part.diceResults.reduce((sum, each) => sum + each) * sign;
        } else if (part.modified === "advantage") {
          return Math.max(...part.diceResults) * sign;
        } else if (part.modified === "disadvantage") {
          return Math.min(...part.diceResults) * sign;
        }
        assertNever(part.modified);
      }
      assertNever(part);
    })
    .reduce((sum, each) => sum + each);
}
