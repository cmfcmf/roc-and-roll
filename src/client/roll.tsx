import React from "react";
import { randomBetweenInclusive } from "../shared/random";
import {
  RRMultipleRoll,
  RRLogEntryDiceRoll,
  RRDamageType,
  RRDice,
  RRPlayerID,
  RRModifier,
} from "../shared/state";
import { assertNever } from "../shared/util";

export function rollInitiative(
  mod: number,
  multiple: RRMultipleRoll,
  playerId: RRPlayerID
): Omit<RRLogEntryDiceRoll, "id" | "type" | "timestamp"> {
  return {
    payload: {
      rollType: "initiative",
      dice: [rollD20(multiple), ...(mod === 0 ? [] : [modifier(mod)])],
    },
    silent: false,
    playerId,
  };
}

function modifier(
  modifier: number,
  damageType: RRDamageType = null
): RRModifier {
  return {
    type: "modifier",
    modifier,
    damageType,
  };
}

function rollD20(multiple: RRMultipleRoll = "none") {
  return roll({
    faces: 20,
    count: multiple == "none" ? 1 : 2,
    modified: multiple,
  });
}

export function roll({
  faces,
  count,
  damageType,
  modified,
  negated,
}: {
  faces: number;
  count: number;
  damageType?: RRDamageType;
  modified?: RRMultipleRoll;
  negated?: boolean;
}): RRDice {
  if (modified !== "none" && count <= 1) {
    count = 2;
  }
  const results = Array.from({ length: count }, () =>
    randomBetweenInclusive(1, faces)
  );
  return {
    type: "dice",
    faces,
    modified: modified ?? "none",
    diceResults: results,
    damageType: damageType ?? null,
    negated: negated ?? false,
  };
}

export function diceResultString(logEntry: RRLogEntryDiceRoll) {
  return logEntry.payload.dice
    .map((part) => {
      if (part.type === "modifier") {
        return <b>{part.modifier}</b>;
      }
      const prefix = part.negated ? "-" : "";
      if (part.type === "dice") {
        if (part.diceResults.length === 1) {
          return (
            <>
              {prefix}
              <b>{part.diceResults[0]!}</b> (d{part.faces})
            </>
          );
        }
        if (part.modified === "none") {
          return (
            <>
              {prefix}
              {part.diceResults
                .map((r) => (
                  <>
                    <b>{r}</b> (d{part.faces})
                  </>
                ))
                .reduce((prev, curr, index) => {
                  return (
                    <>
                      {prev} + {curr}
                    </>
                  );
                })}
            </>
          );
        }
        const boldValue =
          part.modified === "advantage"
            ? Math.max(...part.diceResults)
            : Math.min(...part.diceResults);
        return (
          <>
            {prefix}
            {part.modified === "advantage" ? "a" : "i"}(
            {part.diceResults
              .map((r) => {
                if (r === boldValue) {
                  return (
                    <>
                      <b>{r}</b> (d{part.faces})
                    </>
                  );
                } else {
                  return (
                    <>
                      {r} (d{part.faces})
                    </>
                  );
                }
              })
              .reduce((prev, curr, index) => {
                return (
                  <>
                    {prev}, {curr}
                  </>
                );
              })}
            )
          </>
        );
      }
      assertNever(part);
    })
    .reduce((prev, curr, index) => {
      return (
        <>
          {prev} + {curr}
        </>
      );
    });
}

export function diceResult(logEntry: RRLogEntryDiceRoll) {
  return logEntry.payload.dice
    .map((part) => {
      if (part.type === "modifier") {
        return part.modifier;
      }
      if (part.type === "dice") {
        const sign = part.negated ? -1 : 1;
        if (part.modified === "none") {
          return part.diceResults.reduce((sum, each) => sum + each) * sign;
        } else if (part.modified === "advantage") {
          return Math.max(...part.diceResults) * sign;
        } else if (part.modified === "disadvantage") {
          return Math.min(...part.diceResults) * sign;
        }
        assertNever(part.modified);
      }
      assertNever(part);
    })
    .reduce((sum, each) => sum + each);
}
