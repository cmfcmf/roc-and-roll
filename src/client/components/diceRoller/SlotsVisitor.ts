import {
  DRTPartNum,
  DRTPartDice,
  DRTPartTerm,
  DRTPartParens,
  DRTPartNegated,
  RRDamageType,
} from "../../../shared/dice-roll-tree-types-and-validation";
import { colorForDamageType } from "../../../shared/state";
import { DRTVisitor } from "../../dice-rolling/grammar";

interface DisplayPart {
  damageType: RRDamageType;
}
interface DisplayModifier extends DisplayPart {
  type: "modifier";
  modifier: number;
}
interface DisplayDie extends DisplayPart {
  type: "die";
  faces: number;
  result: number;
  used: boolean;
  color: string;
}
interface DisplayWeirdDie extends DisplayPart {
  type: "weirdDie";
  faces: number;
  result: number;
  used: boolean;
  color: string;
}

export type RollSlots = (DisplayDie | DisplayModifier | DisplayWeirdDie)[];

export class SlotsVisitor extends DRTVisitor<RollSlots, true, boolean> {
  protected override visitNum(
    expression: DRTPartNum,
    negated?: boolean
  ): RollSlots {
    return [
      {
        type: "modifier",
        modifier: (negated ? -1 : 1) * expression.value,
        damageType: expression.damage,
      },
    ];
  }

  protected override visitDice(expression: DRTPartDice<true>): RollSlots {
    return expression.results.map((result) => ({
      type: [4, 6, 8, 10, 12, 20].includes(expression.faces)
        ? "die"
        : "weirdDie",
      damageType: expression.damage,
      faces: expression.faces,
      result: result,
      color:
        expression.faces !== 20
          ? colorForDamageType(expression.damage.type)
          : result === 1
          ? "darkred"
          : result === 20
          ? "green"
          : "orange",
      used:
        expression.modified === "none"
          ? true
          : expression.modified === "advantage"
          ? result === Math.max(...expression.results)
          : result === Math.min(...expression.results),
    }));
  }

  protected override visitTerm(
    expression: DRTPartTerm<true>,
    negated?: boolean
  ): RollSlots {
    return expression.operands.flatMap((operand, i) =>
      this.visit(
        operand,
        expression.operator === "-" && i > 0 ? !(negated ?? false) : negated
      )
    );
  }

  protected override visitParens(
    expression: DRTPartParens<true>,
    negated?: boolean
  ): RollSlots {
    return this.visit(expression.inner, negated);
  }

  protected override visitNegated(
    expression: DRTPartNegated<true>,
    negated?: boolean
  ): RollSlots {
    return this.visit(expression.inner, !(negated ?? false));
  }
}
