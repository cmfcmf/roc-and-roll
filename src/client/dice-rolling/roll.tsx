import React from "react";
import {
  DiceRollTree,
  DRTPartDice,
  DRTPartExpression,
  DRTPartNegated,
  DRTPartNum,
  DRTPartParens,
  DRTPartTerm,
  RRDamageType,
} from "../../shared/dice-roll-tree-types-and-validation";
import { randomBetweenInclusive } from "../../shared/random";
import {
  RRMultipleRoll,
  RRLogEntryDiceRoll,
  RRPlayerID,
  colorForDamageType,
  damageTypes,
} from "../../shared/state";
import { assertNever } from "../../shared/util";
import { contrastColor } from "../util";
import { DRTVisitor, parseDiceString } from "./grammar";

export function isValidDiceString(diceString: string) {
  try {
    parseDiceString(diceString);
    return true;
  } catch (err) {
    return false;
  }
}

export class RollVisitor extends DRTVisitor<DRTPartExpression<true>, false> {
  // You can overwrite this function in tests to roll deterministically.
  static rollFunction = randomBetweenInclusive;

  protected override visitNum(expression: DRTPartNum) {
    return expression;
  }

  protected override visitDice(expression: DRTPartDice<false>) {
    return {
      ...expression,
      results: Array.from({ length: expression.count }, () =>
        RollVisitor.rollFunction(1, expression.faces)
      ),
    };
  }

  protected override visitTerm(expression: DRTPartTerm<false>) {
    return {
      ...expression,
      operands: expression.operands.map((operand) => this.visit(operand)),
    };
  }

  protected override visitParens(expression: DRTPartParens<false>) {
    return {
      ...expression,
      inner: this.visit(expression.inner),
    };
  }

  protected override visitNegated(expression: DRTPartNegated<false>) {
    return {
      ...expression,
      inner: this.visit(expression.inner),
    };
  }
}

export function parseDiceStringAndRoll(diceString: string) {
  return rollDiceRollTree(parseDiceString(diceString));
}

export function rollDiceRollTree(diceRollTree: DiceRollTree<false>) {
  return new RollVisitor().visit(diceRollTree);
}

export function rollInitiative(
  modifier: number,
  multiple: RRMultipleRoll,
  playerId: RRPlayerID
): Omit<RRLogEntryDiceRoll, "id" | "type" | "timestamp"> {
  const d20 = rollD20(multiple);

  return {
    payload: {
      rollType: "initiative",
      rollName: "Intiative",
      diceRollTree:
        modifier === 0
          ? d20
          : {
              type: "term",
              operator: "+",
              operands: [
                d20,
                {
                  type: "num" as const,
                  value: modifier,
                  damage: {
                    type: null,
                  },
                },
              ],
            },
    },
    silent: false,
    playerId,
  };
}

function rollD20(multiple: RRMultipleRoll = "none"): DRTPartDice<true> {
  return roll({
    faces: 20,
    count: multiple === "none" ? 1 : 2,
    modified: multiple,
    damage: {
      type: null,
    },
  });
}

export function roll({
  faces,
  count,
  damage,
  modified,
}: {
  faces: number;
  count: number;
  damage: RRDamageType;
  modified?: RRMultipleRoll;
}): DRTPartDice<true> {
  if (modified !== undefined && modified !== "none" && count <= 1) {
    count = 2;
  }
  const results = Array.from({ length: count }, () =>
    randomBetweenInclusive(1, faces)
  );
  return {
    type: "dice",
    faces,
    modified: modified ?? "none",
    count,
    results,
    damage,
  };
}

class FormatDiceRollVisitor extends DRTVisitor<React.ReactElement, true> {
  protected override visitNum(expression: DRTPartNum) {
    return <b>{expression.value}</b>;
  }

  protected override visitDice(expression: DRTPartDice<true>) {
    if (expression.results.length === 1) {
      return (
        <>
          <b>{expression.results[0]!}</b> (d{expression.faces})
        </>
      );
    }
    if (expression.modified === "none") {
      return (
        <>
          <b>
            [{expression.results.join(", ")}] ({expression.count}d
            {expression.faces})
          </b>
        </>
      );
    }
    const boldValue =
      expression.modified === "advantage"
        ? Math.max(...expression.results)
        : Math.min(...expression.results);
    return (
      <>
        {expression.modified === "advantage" ? "a" : "i"}(
        {expression.results
          .map((r) => {
            if (r === boldValue) {
              return (
                <>
                  <b>{r}</b> (d{expression.faces})
                </>
              );
            } else {
              return (
                <>
                  {r} (d{expression.faces})
                </>
              );
            }
          })
          .reduce((prev, curr) => {
            return (
              <>
                {prev}, {curr}
              </>
            );
          })}
        )
      </>
    );
  }

  protected override visitTerm(expression: DRTPartTerm<true>) {
    return expression.operands
      .map((operand) => this.visit(operand))
      .reduce((prev, curr) => {
        return (
          <>
            {prev} {expression.operator} {curr}
          </>
        );
      });
  }

  protected override visitParens(expression: DRTPartParens<true>) {
    return <>({this.visit(expression.inner)})</>;
  }

  protected override visitNegated(expression: DRTPartNegated<true>) {
    return <>-({this.visit(expression.inner)})</>;
  }
}

export function diceResultString(diceRollTree: DiceRollTree<true>) {
  return new FormatDiceRollVisitor().visit(diceRollTree);
}

class SumVisitor extends DRTVisitor<number, true> {
  protected override visitNum(expression: DRTPartNum) {
    return expression.value;
  }

  protected override visitDice(expression: DRTPartDice<true>) {
    switch (expression.modified) {
      case "none":
        return expression.results.reduce((sum, each) => sum + each);
      case "advantage":
        return Math.max(...expression.results);
      case "disadvantage":
        return Math.min(...expression.results);
      default:
    }
    assertNever(expression.modified);
  }

  protected override visitTerm(expression: DRTPartTerm<true>) {
    return expression.operands
      .map((operand) => this.visit(operand))
      .reduce((a, b) => {
        switch (expression.operator) {
          case "+":
            return a + b;
          case "-":
            return a - b;
          case "*":
            return a * b;
          case "/":
            return a / b;
          default:
            assertNever(expression.operator);
        }
      });
  }

  protected override visitParens(expression: DRTPartParens<true>) {
    return this.visit(expression.inner);
  }

  protected override visitNegated(expression: DRTPartNegated<true>) {
    return -this.visit(expression.inner);
  }
}

export function diceResult(diceRollTree: DiceRollTree<true>) {
  return new SumVisitor().visit(diceRollTree);
}

type DamageSumMap = Record<
  NonNullable<RRDamageType["type"]> | "unspecified",
  number
>;

class SumWithDamageTypeVisitor extends DRTVisitor<DamageSumMap, true> {
  private makeMap(): DamageSumMap {
    return {
      unspecified: 0,
      slashing: 0,
      piercing: 0,
      bludgeoning: 0,
      poison: 0,
      acid: 0,
      fire: 0,
      cold: 0,
      radiant: 0,
      necrotic: 0,
      lightning: 0,
      thunder: 0,
      force: 0,
      psychic: 0,
    };
  }

  private op(
    fn: (a: number, b: number) => number,
    a: DamageSumMap,
    b: DamageSumMap
  ) {
    const result = this.makeMap();
    for (const damageType of damageTypes) {
      result[damageType ?? "unspecified"] = fn(
        a[damageType ?? "unspecified"],
        b[damageType ?? "unspecified"]
      );
    }
    return result;
  }

  protected override visitNum(expression: DRTPartNum): DamageSumMap {
    const damageType = expression.damage.type ?? "unspecified";

    return {
      ...this.makeMap(),
      [damageType]: expression.value,
    };
  }

  protected override visitDice(expression: DRTPartDice<true>): DamageSumMap {
    const damageType = expression.damage.type ?? "unspecified";

    switch (expression.modified) {
      case "none":
        return {
          ...this.makeMap(),
          [damageType]: expression.results.reduce((sum, each) => sum + each),
        };
      case "advantage":
        return {
          ...this.makeMap(),
          [damageType]: Math.max(...expression.results),
        };
      case "disadvantage":
        return {
          ...this.makeMap(),
          [damageType]: Math.min(...expression.results),
        };
      default:
    }
    assertNever(expression.modified);
  }

  protected override visitTerm(expression: DRTPartTerm<true>): DamageSumMap {
    return expression.operands
      .map((operand) => this.visit(operand))
      .reduce((a, b) =>
        this.op(
          (a, b) => {
            switch (expression.operator) {
              case "+":
                return a + b;
              case "-":
                return a - b;
              case "*":
                return a * b;
              case "/":
                return a / b;
              default:
                assertNever(expression.operator);
            }
          },
          a,
          b
        )
      );
  }

  protected override visitParens(
    expression: DRTPartParens<true>
  ): DamageSumMap {
    return this.visit(expression.inner);
  }

  protected override visitNegated(
    expression: DRTPartNegated<true>
  ): DamageSumMap {
    return this.op(
      (damage) => -damage,
      this.visit(expression.inner),
      this.makeMap()
    );
  }
}

export function DiceResultWithTypes({
  logEntry,
}: {
  logEntry: RRLogEntryDiceRoll;
}) {
  const total = diceResult(logEntry.payload.diceRollTree);
  const damageTypeSums = new SumWithDamageTypeVisitor().visit(
    logEntry.payload.diceRollTree
  );

  function onlyHasUnspecifiedDamage(damageTypeSums: DamageSumMap) {
    for (const damageType of damageTypes) {
      if (damageType === null) {
        continue;
      }
      if (damageTypeSums[damageType] !== 0) {
        return false;
      }
    }
    return true;
  }

  return (
    <>
      <strong>{total}</strong>
      {!onlyHasUnspecifiedDamage(damageTypeSums) && (
        <>
          {" ("}
          {(
            Object.entries(damageTypeSums) as Array<
              [keyof DamageSumMap, number]
            >
          ).map(([type, sum]) => {
            if (sum === 0 || isNaN(sum)) {
              // Division by 0 may result in NaN.
              return null;
            }
            const color = colorForDamageType(
              type === "unspecified" ? null : type
            );
            return (
              <span
                key={type}
                className="dice-results-per-damage"
                style={{
                  backgroundColor: color,
                  color: contrastColor(color),
                }}
                title={type}
              >
                {sum}
              </span>
            );
          })}
          )
        </>
      )}
    </>
  );
}
