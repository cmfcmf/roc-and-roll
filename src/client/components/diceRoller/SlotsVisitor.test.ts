import { parseDiceStringAndRoll } from "../../dice-rolling/roll";
import { SlotsVisitor } from "./SlotsVisitor";

describe("SlotsVisitor", () => {
  it.each(["-5", "--5", "-(5)", "10-(5)", "10+-5", "10-5", "1-(2-3)"])(
    "correctly handles negative numbers in all cases %s",
    (diceRoll: string) => {
      const visitor = new SlotsVisitor();
      const result = visitor.visit(parseDiceStringAndRoll(diceRoll));
      expect(result).toMatchSnapshot();
    }
  );
});
