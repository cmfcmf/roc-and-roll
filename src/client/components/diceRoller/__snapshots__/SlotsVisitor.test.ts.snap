// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlotsVisitor correctly handles negative numbers in all cases -(5) 1`] = `
Array [
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": -5,
    "type": "modifier",
  },
]
`;

exports[`SlotsVisitor correctly handles negative numbers in all cases --5 1`] = `
Array [
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": 5,
    "type": "modifier",
  },
]
`;

exports[`SlotsVisitor correctly handles negative numbers in all cases -5 1`] = `
Array [
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": -5,
    "type": "modifier",
  },
]
`;

exports[`SlotsVisitor correctly handles negative numbers in all cases 1-(2-3) 1`] = `
Array [
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": 1,
    "type": "modifier",
  },
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": -2,
    "type": "modifier",
  },
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": 3,
    "type": "modifier",
  },
]
`;

exports[`SlotsVisitor correctly handles negative numbers in all cases 10+-5 1`] = `
Array [
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": 10,
    "type": "modifier",
  },
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": -5,
    "type": "modifier",
  },
]
`;

exports[`SlotsVisitor correctly handles negative numbers in all cases 10-(5) 1`] = `
Array [
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": 10,
    "type": "modifier",
  },
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": -5,
    "type": "modifier",
  },
]
`;

exports[`SlotsVisitor correctly handles negative numbers in all cases 10-5 1`] = `
Array [
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": 10,
    "type": "modifier",
  },
  Object {
    "damageType": Object {
      "type": null,
    },
    "modifier": -5,
    "type": "modifier",
  },
]
`;
